Improve implementation of findClusterIndex and findClusterValue, add test

mfridental opened the pull request:
Sorry, while writing a test I have found out that the current implementation was wrong. Here is a fix and a test.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

mfridental committed:
Improve implementation of findClusterIndex and findClusterValue, add test

robot-metrika-test commented:
Can one of the admins verify this patch?

alexey-milovidov commented on dbms/src/Functions/FunctionsFindCluster.h:
```diff
@@ -270,35 +142,129 @@ class FunctionFindClusterIndex : public IFunction
         return ClusterOperation::FindClusterIndex;
     }
 
+    virtual void executeImplTyped(
```
Function should not be virtual, unless you are going to override it and then to use the object in polymorphic way.

alexey-milovidov commented on dbms/src/Functions/FunctionsFindCluster.h:
```diff
@@ -325,6 +291,30 @@ class FunctionFindClusterValue : public FunctionFindClusterIndex
         return ClusterOperation::FindCentroidValue;
     }
+
+    virtual void executeImplTyped(
```
Same.

alexey-milovidov commented on dbms/src/Functions/FunctionsFindCluster.h:
```diff
@@ -325,6 +291,30 @@ class FunctionFindClusterValue : public FunctionFindClusterIndex
+
+    virtual void executeImplTyped(
+    		const IColumn* in_untyped,
+			IColumn* out_untyped,
```
Tabs and spaces are messed up.
We are using indentation by four spaces.

alexey-milovidov reviewed (changes requested):
Great. Just a few changes needed.

mfridental committed:
findClusterIndex: fix virtual method, improve formatting

mfridental commented:
executeImplTyped is intended to be virtual; added "override" in the child class.
Reformatted using a code style without tabs. 
It would be nice if you share eclipse settings or at least formatting style to all contributors. Ideally, a whole development environment in a VM.

By the way, in my environment the test 00107 crashes the server, already since many weeks. I suppose its just my environment?

alexey-milovidov commented:
> It would be nice if you share eclipse settings or at least formatting style to all contributors. Ideally, a whole development environment in a VM.

You could try to use `clang-format` for new or rewritten files.
(For this you should have latest clang (4.0). Use it as following: `clang-format -style=file -i filename.cpp`)

We have much variety of what environment our developers prefer.

> By the way, in my environment the test 00107 crashes the server, already since many weeks. I suppose its just my environment?

No idea. This test passes in our CI on Linux with ClickHouse compiled by gcc-6 or clang-4. Also tested on clang-4 with ASan and gcc-6 under Valgrind. This test has been always passed in recent time.
What is your environment: OS, machine, compiler? Do you use latest master or stable tag to base your branch (both should be Ok)?

mfridental commented:
Here is my environment:

immowelt@ubuntu:~/ClickHouse$ uname -r
4.8.0-36-generic

immowelt@ubuntu:~/ClickHouse$ lsb_release -a
No LSB modules are available.
Distributor ID:    Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:    16.04
Codename:    xenial

immowelt@ubuntu:~/ClickHouse$ cat /proc/cpuinfo
processor    : 0
vendor_id    : GenuineIntel
cpu family    : 6
model        : 94
model name    : Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz
stepping    : 3
microcode    : 0x19
cpu MHz        : 3200.110
cache size    : 6144 KB
physical id    : 0
siblings    : 3
core id        : 0
cpu cores    : 3
apicid        : 0
initial apicid    : 0
fpu        : yes
fpu_exception    : yes
cpuid level    : 22
wp        : yes
flags        : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch epb fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 invpcid rtm rdseed adx smap xsaveopt dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs        :
bogomips    : 6402.00
clflush size    : 64
cache_alignment    : 64
address sizes    : 42 bits physical, 48 bits virtual
power management:
processor    : 1
vendor_id    : GenuineIntel
cpu family    : 6
model        : 94
model name    : Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz
stepping    : 3
microcode    : 0x19
cpu MHz        : 3200.110
cache size    : 6144 KB
physical id    : 0
siblings    : 3
core id        : 1
cpu cores    : 3
apicid        : 1
initial apicid    : 1
fpu        : yes
fpu_exception    : yes
cpuid level    : 22
wp        : yes
flags        : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch epb fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 invpcid rtm rdseed adx smap xsaveopt dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs        :
bogomips    : 6402.00
clflush size    : 64
cache_alignment    : 64
address sizes    : 42 bits physical, 48 bits virtual
power management:
processor    : 2
vendor_id    : GenuineIntel
cpu family    : 6
model        : 94
model name    : Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz
stepping    : 3
microcode    : 0x19
cpu MHz        : 3200.110
cache size    : 6144 KB
physical id    : 0
siblings    : 3
core id        : 2
cpu cores    : 3
apicid        : 2
initial apicid    : 2
fpu        : yes
fpu_exception    : yes
cpuid level    : 22
wp        : yes
flags        : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch epb fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 invpcid rtm rdseed adx smap xsaveopt dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
bugs        :
bogomips    : 6402.00
clflush size    : 64
cache_alignment    : 64
address sizes    : 42 bits physical, 48 bits virtual
power management:

immowelt@ubuntu:~/ClickHouse$ gcc --version
gcc (Ubuntu 5.4.1-2ubuntu1~16.04) 5.4.1 20160904
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

immowelt@ubuntu:~/ClickHouse$ git rev-parse HEAD
099482ad10163c3fcc0a3d81f509c62fca5b80e4
(based on recent master branch)


The crash happens here: dbms/src/Server/clickhouse(DB::ColumnAggregateFunction::~ColumnAggregateFunction())

After adding tracing to this destructor:
ColumnAggregateFunction::~ColumnAggregateFunction()
{
    LOG_DEBUG((&Logger::get("crash")), "enter");
    LOG_DEBUG((&Logger::get("crash")), func->getName());

    if (!func->hasTrivialDestructor() && !src)
    {
        LOG_DEBUG((&Logger::get("crash")), "if");

        for (auto val : data)
        {
            LOG_DEBUG((&Logger::get("crash")), std::to_string(UInt64(val)));
            func->destroy(val);
        }
    }
}

I see the following printout:

2017.07.14 23:15:27.411902 [ 6 ] <Trace> Aggregator: Max size of temporary block: 391 rows, 32.009 MiB.
2017.07.14 23:15:27.414229 [ 6 ] <Trace> Aggregator: Written part in 1.212 sec., 100001 rows, 2.748 MiB uncompressed, 1.536 MiB compressed, 28.813 uncompressed bytes per row, 16.106 compressed bytes per row, compression rate: 1.789 (82483.756 rows/sec., 2.267 MiB/sec. uncompressed, 1.267 MiB/sec. compressed)
2017.07.14 23:15:27.423301 [ 6 ] <Trace> Aggregator: Aggregated. 500000 to 1 rows (from 14.676 MiB) in 1.249 sec. (400169.372 rows/sec., 11.746 MiB/sec.)
2017.07.14 23:15:27.423470 [ 6 ] <Trace> AggregatingBlockInputStream: Will merge 1 temporary files of size 1.53602 MiB compressed, 2.74786 MiB uncompressed.
2017.07.14 23:15:27.423510 [ 6 ] <Trace> Aggregator: Destroying aggregate states
2017.07.14 23:15:27.424491 [ 6 ] <Trace> Aggregator: Merging partially aggregated blocks (bucket = 0).
2017.07.14 23:15:27.424604 [ 6 ] <Debug> crash: enter
2017.07.14 23:15:27.424670 [ 6 ] <Debug> crash: count
2017.07.14 23:15:27.424690 [ 6 ] <Debug> crash: enter
2017.07.14 23:15:27.424708 [ 6 ] <Debug> crash: argMax
2017.07.14 23:15:27.424722 [ 6 ] <Debug> crash: if
2017.07.14 23:15:27.424729 [ 6 ] <Debug> crash: 156979200
2017.07.14 23:15:27.424742 [ 6 ] <Debug> crash: 156979280
2017.07.14 23:15:27.424746 [ 6 ] <Debug> crash: 156979360
[...]
2017.07.14 23:15:27.428352 [ 6 ] <Debug> crash: 183478384
2017.07.14 23:15:27.428356 [ 6 ] <Debug> crash: 183478464
2017.07.14 23:15:27.428369 [ 6 ] <Debug> crash: 183478544
2017.07.14 23:15:27.428373 [ 6 ] <Debug> crash: 183478624
2017.07.14 23:15:27.428386 [ 6 ] <Debug> crash: 183478704
2017.07.14 23:15:27.428390 [ 6 ] <Debug> crash: 183478784
2017.07.14 23:15:27.428403 [ 6 ] <Debug> crash: 183478864
2017.07.14 23:15:27.428411 [ 6 ] <Trace> Aggregator: Converting aggregated data to blocks
2017.07.14 23:15:27.428448 [ 6 ] <Trace> Aggregator: Converted aggregated data to blocks. 392 rows, 0.259 MiB in 0.000 sec. (22028659.736 rows/sec., 14553.086 MiB/sec.)
2017.07.14 23:15:27.428468 [ 6 ] <Debug> crash: enter
2017.07.14 23:15:27.428471 [ 6 ] <Debug> crash: count
2017.07.14 23:15:27.428489 [ 6 ] <Debug> crash: enter
2017.07.14 23:15:27.428495 [ 6 ] <Debug> crash: argMax
2017.07.14 23:15:27.428508 [ 6 ] <Debug> crash: if
2017.07.14 23:15:27.428511 [ 6 ] <Debug> crash: 8
2017.07.14 23:15:27.428538 [ 7 ] <Error> BaseDaemon: ########################################
2017.07.14 23:15:27.428644 [ 7 ] <Error> BaseDaemon: (from thread 6) Received signal Segmentation fault (11).
2017.07.14 23:15:27.428651 [ 7 ] <Error> BaseDaemon: Address: 0x8
2017.07.14 23:15:27.450564 [ 7 ] <Error> BaseDaemon: 1. dbms/src/Server/clickhouse(DB::ColumnAggregateFunction::~ColumnAggregateFunction()+0x7a9) [0x2af0fc9]

Do you have any idea how this 8 is getting there? Can it be related with serialization of the state?

I’ve installed gcc-6 by using
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt-get install gcc-6 g++-6

removed gcc-5 and recompiled the server – the crash is still there


From: alexey-milovidov 
Sent: Thursday, July 13, 2017 11:51 PM
To: yandex/ClickHouse 
Cc: mfridental ; Author 
Subject: Re: [yandex/ClickHouse] Improve implementation of findClusterIndex and findClusterValue, add test (#980)

  It would be nice if you share eclipse settings or at least formatting style to all contributors. Ideally, a whole development environment in a VM.

You could try to use clang-format for new or rewritten files.
(For this you should have latest clang (4.0). Use it as following: clang-format -style=file -i filename.cpp)

We have much variety of what environment our developers prefer.

  By the way, in my environment the test 00107 crashes the server, already since many weeks. I suppose its just my environment?

No idea. This test passes in our CI on Linux with ClickHouse compiled by gcc-6 or clang-4. Also tested on clang-4 with ASan and gcc-6 under Valgrind. This test has been always passed in recent time.
What is your environment: OS, machine, compiler? Do you use latest master or stable tag to base your branch (both should be Ok)?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.
